### PR TITLE
Support variable array indexing for header stacks

### PR DESCRIPTION
**Pull request by usha1830**
This PR is to support variable array indexing for accessing the elements of a header stack using subscript notation.
The array index expression within the following statements/expressions are supported

- Assignment statements (on both sides as well)
- Table keys
- If condition
- Switch expression
- Function param
- Binary operation
- Unary Operation
- Slice

`meta.left = hdrs.udp[meta.depth - 1].sport;`

should be translated as 

```
tmp = meta.depth - 1
if (tmp == 0)
    meta.left = hdrs.udp[0].sport;
if (tmp == 1)
    meta.left = hdrs.udp[1].sport;
```

If the array subscript is present on both sides with different expression in the [], it will be translated into multi level branches like below:

`hdrs.udp[meta.depth].sport = hdrs.udp[meta.depth - 1].sport;`  =>

```
tmp = meta.depth - 1
tmp1 = meta.depth

if (tmp1 == 0)
    if (tmp == 0)
        hdr.udp[0].sport = hdr.udp[0].sport
if (tmp == 1)
        hdr.udp[0].sport = hdr.udp[1].sport
if (tmp1 == 1)
    if (tmp == 0)
        hdr.udp[1].sport = hdr.udp[0].sport
if (tmp == 1)
        hdr.udp[1].sport = hdr.udp[1].sport		
```

### CODE REVIEW

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -1046,7 +1052,14 @@ bool LogicalExpressionUnroll::preorder(const IR::MethodCallExpression *m) {
...
 }
 
-bool LogicalExpressionUnroll::preorder(const IR::Member *) {
+bool LogicalExpressionUnroll::preorder(const IR::Member *m) {
+    if (m->expr->is<IR::ArrayIndex>()) {
+        auto tmp = new IR::PathExpression(IR::ID(refMap->newName("tmp")));
+        root = tmp;
+        decl.push_back(new IR::Declaration_Variable(tmp->path->name, m->type));
+        stmt.push_back(new IR::AssignmentStatement(root, m));
+        return false;
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.cpp:1061]: any specific reason to have tmp variable? root shall be used.

File: backends/dpdk/dpdkArch.cpp [hunk 2]
```diff
@@ -1397,15 +1410,205 @@ const IR::Node* DismantleMuxExpressions::preorder(IR::Mux* expression) {
...
+void SplitHSIndexExpression::replaceVarIndexWithIf(IR::AssignmentStatement *statement,
+                                                    bool leftHasHSE, bool rightHasHSE) {
+    const IR::Member *mem = nullptr;
+    const IR::ArrayIndex *ai = nullptr;
+    size_t n_elem = 0;
+    if (leftHasHSE && rightHasHSE) {
+        mem = statement->left->to<IR::Member>();
+        if (mem->expr->type->is<IR::Type_Stack>())
+
+        ai = mem->expr->to<IR::ArrayIndex>();
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.cpp:1446]: not indented properly, a blank also there.

File: backends/dpdk/dpdkArch.cpp [hunk 3]
```diff
@@ -1397,15 +1410,205 @@ const IR::Node* DismantleMuxExpressions::preorder(IR::Mux* expression) {
...
+        if (ai->left->is<IR::Member>())
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.cpp]: ai shall be nullptr if 'mem->expr->type->is<IR::Type_Stack>()' is false.

**Review by psivanup — CHANGES_REQUESTED** [review]
Please check the comments.

File: backends/dpdk/dpdkArch.cpp (hunk 1 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:1061]: It caused some issue. I am trying to recollect why I introduced tmp here :(

Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp:1061]: In this pass, root is not PathExpression type, it is of type IR::Expression. directly assigning with new path expression throws build error.

Comment (psivanup) [reply backends/dpdk/dpdkArch.cpp:1061]: Ok

File: backends/dpdk/dpdkArch.cpp (hunk 3 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp]: Updated to move the hasHSE check from Assignment statement visitor to replaceVarIndexWithIf function. This makes sure the pointers are not null.
Please check.

File: backends/dpdk/dpdkArch.cpp [hunk 4]
```diff
@@ -1397,15 +1410,205 @@ const IR::Node* DismantleMuxExpressions::preorder(IR::Mux* expression) {
...
+
+
+const IR::Node* PrepSplitHSIndexExpression::preorder(IR::Key* keys) {
+    LOG3("Visiting " << keys);
+    bool complex = false;
+    for (auto k : keys->keyElements) {
+         if (auto keyField = k->expression->to<IR::Member>()) {
+             if (keyField->expr->is<IR::ArrayIndex>() &&
+                 keyField->expr->type->is<IR::Type_Header>()) {
+                 complex = true;
```
Comment (psivanup) [inline backends/dpdk/dpdkArch.cpp:1562]: can add an early break here.

**Review by psivanup — APPROVED** [review]
Looks ok to me, there are optional comments though.

File: backends/dpdk/dpdkArch.cpp [hunk 5]
```diff
@@ -1397,15 +1410,177 @@ const IR::Node* DismantleMuxExpressions::preorder(IR::Mux* expression) {
...
+    replaceVarIndexWithIf(statement);
+    if (statements.empty())
+        return statement;
+    auto block = new IR::BlockStatement(statements);
+    statements.clear();
+    return block;
+}
+
+const IR::Node* PrepSplitHSIndexExpression::preorder(IR::StructField *sf) {
+    if (sf->type->is<IR::Type_Stack>()){
```
Comment (github-sajan) [inline backends/dpdk/dpdkArch.cpp]: cpplint error ? space missing before curly braces

**Review by github-sajan — APPROVED** [review]
Code looks fine to me

**Review by kamleshbhalui — APPROVED** [review]

File: backends/dpdk/dpdkArch.cpp (hunk 5 above)
Comment (usha1830) [reply backends/dpdk/dpdkArch.cpp]: Will update

**Review by psivanup — APPROVED** [review]
The changes look fine to me.

**Pull request closed without merging**